src/fauxtoshop.cpp

```diff
@@ -1,517 +1,487 @@
+ * return the same sequence, and use row-major order to generate 'dr' and 'dc' values
+ * Params: original - image Grid of integers
+ * ------------------------------------------------------------------------------------------------
+ * Returns: nothing. Void function
+*/
+void applyScatterFilter(GBufferedImage &img){
+    // Convert your GBufferedImage object into a Grid<int>
+    Grid<int> original = img.toGrid();
+    Grid<int> scatterGrid(original.numRows(), original.numCols());
+    // Prompt user for scatter radius input between 1 and 100 (inclusive)
+    int scatterRadius = getValidInteger(
+                "Enter the scatter radius as number between 1 and 100: ", 1, 100);
+    // Iterate through each pixel in the image grid
+    for (int i = 0; i < scatterGrid.numRows(); i++){
+        for (int j = 0; j < scatterGrid.numCols(); j++) {
+        // Randamly select nearby pixel (row and col) to provide new colour
+        getRandomPixelColor(original, scatterGrid, i, j, scatterRadius);
+        }
+    }
+    img.fromGrid(scatterGrid);
+}
+
+/*
+ * Function: applyEdgeDetectionFilter
+ * Usage: Create a new black and white image the same size as the original. A pixel is black if it
+ * is an edge or white if it is not an edge
+ * Function:
+ * a. Prompt user to provide an edge threshold that specifies the difference between 2 pixels
+ * (positive integer) before they make an 'edge'. If input is invalid reprompt
+ * b. Loop over each pixel to determine whether it is an edge or not
+ * c. Check all 9 pixels (including self at 0) to check if at least 1 neighbour has a difference
+ * greater than the threshold e.g. the pixel is an edge for [9, 8, 5, 3, 3, 0 (self), 4, 7, 8, 7]
+ * if threshold < 9
+ * Hint: pixels near edges and corners may not have all 9 neighbors! Use Grid class inBounds member
+ * function to cater for this
+ * Params: original - image Grid of integers, blackWhiteGrid - image Grid of integers
+ * ------------------------------------------------------------------------------------------------
+ * Returns: nothing. Void function
+*/
+void applyEdgeDetectionFilter(GBufferedImage &img){
+    // Prompt user for scatter radius input as a positive integer
+    int edgeThreshold = getValidInteger(
+                "Enter the edge threshold as a positive number: ", 0, INT_MAX);
+    // Set up a new image grid for the black and white output for edge detection
+    Grid<int> original = img.toGrid();
+    Grid<int> blackWhiteGrid(original.numRows(), original.numCols());
+    // Iterate through each pixel
+    for (int i = 0; i < original.numRows(); i++){
+        for (int j = 0; j < original.numCols(); j++) {
+            // Define boolean to set whether each pixel is an edge or not
+            bool isEdge = false;
+            // Check each pixel bordering the pixel at relative location (0,0) to detect the edge
+            for (int rowOffset = -1; rowOffset < 2; rowOffset++){
```

Comment:
Recommended: you can extract some of this logic to a helper function. E.g. the helper function can return true if the pixel is an edge, or it can return the color that the pixel should be (i.e. whether it's black or white).

Putting this logic in a helper function will also make this consistent with the level of decomposition you used in your other filter methods.